### ISSUE TITLE
findById...AndXXX method derived query generates wrong statement

### ISSUE THREAD

**Issue by mmonti** [spring-projects/spring-data-couchbase#1072]
I am looking at some Spring Data Couchbase integration [issues](https://github.com/jhipster/generator-jhipster/issues/13544) in JHIpster project, and one particular issue I found is that when we have a repository method to find by id and any other property, the generated query is wrong.

As an example, the method ```findByIdIsNotNullAndActivatedIsTrue``` generates the following statement:
```
SELECT META(`testBucket`).id AS __id, META(`testBucket`).cas AS __cas, `testBucket`.* FROM `testBucket` WHERE `_class` = "tech.jhipster.sample.domain.User" AND id is not null and `activated`
```
In this scenario the Id should have meta(`<bucket-name>`) prepended:
```
SELECT META(`testBucket`).id AS __id, META(`testBucket`).cas AS __cas, `testBucket`.* FROM `testBucket` WHERE `_class` = "tech.jhipster.sample.domain.User" AND META(`testBucket`).`id` is not null and `activated`
```

To support these type of query, I believe that ```N1qlQueryCreator.java``` and ```QueryCriteria.java``` classes needs to be modified. 
QueryCriteria checks for **backtics** at start or end: 
- https://github.com/spring-projects/spring-data-couchbase/blob/master/src/main/java/org/springframework/data/couchbase/core/query/QueryCriteria.java#L449 

and if its not changed, it ends up adding backtics to the META: ``` `META(`testBucket`).`id`` ```)

I did an attempt to support this type of queries here:  

- https://github.com/mraible/ngx-couchbase/pull/1/files#diff-12436df3f4d579c807f608e52f5ad8d1a924c369023cdfcdc03b146b2b11d76eR64-R73 

and even though it works, I am not sure if its the right place for that change. 
I can probably create a PR with the fix if needed.

**Comment by mikereiche** [comment]
Thanks @mmonti. And thanks for your change-set, I'll take a look at it.

There was another recent issue about the back-ticks around META expressions. https://github.com/spring-projects/spring-data-couchbase/issues/1066 .   A quick-and-dirty work-around is

	private String maybeBackTic(String value) {
		if (value == null || (value.startsWith("`") && value.endsWith("`"))) {
			return value;
		} else if (value == null || (value.contains("META("))) {
			return value;
		} else {
			return "`" + value + "`";
		}
	}

**Comment by mikereiche** [comment]
One thought is to handle this in the NqlQueryCreator Converter.  It could be extended to handle other meta types (cas, expiration).

```	static Converter<? super CouchbasePersistentProperty, String> cvtr = new MyConverter();

	static class MyConverter implements Converter<CouchbasePersistentProperty, String>{
		@Override
		public String convert(CouchbasePersistentProperty source) {
			if(source.isIdProperty()) {
				return "META().id";
			}
			return new StringBuilder(source.getFieldName().length() + 2).append('`').append(source.getFieldName()).append('`').toString();
		}
	}

**Comment by mmonti** [comment]
Ohh wow... that is pretty much what I did... :) One difference is that I modified the N1qlQueryCreator to receive CouchbaseOperations cause I thought you need the bucket name. Don't u need to specify the bucket in ```META(<bucketName>)```??
I commented that out in this commit cause I wasn't sure if it was the right place to put it: 
- https://github.com/mraible/ngx-couchbase/pull/1/commits/0e49bb57bbcebae9f54d2c6ab236cf135fbe857b#diff-13c5eb982491c91dbac8f8731fe426bd0b6e897b82cc6393864203db6fa53b06R144

As for the change in QueryCriteria, the change that I did is a bit different. I think that just checking if the value contains "`" should do it... unless i am missing some specific scenario
```
private String maybeBackTic(String value) {
   if (value == null || (value.startsWith("`") && value.endsWith("`")) || value.contains("`")) {
     return value;
   } else {
     return "`" + value + "`";
   }
}
```

Anyways, Ill prepare a PR and let you know

**Comment by aaronjwhiteside** [comment]
Backticks can be used in identifiers, and but must be escaped.

https://docs.couchbase.com/server/current/n1ql/n1ql-language-reference/identifiers.html#escaped-identifiers

So it may not be a great idea to just check if the value already contains a backtick, as it might be an escaped backtick.

**Comment by mikereiche** [comment]
> Don't u need to specify the bucket in META(<bucketName>)??

Yeah, I kinda punted on that.  You don't need bucketName if the intended bucketName is unambiguous from the rest of the query.

**Comment by mmonti** [comment]
Hi @mikereiche,

I have a branch with changes to fix these two issues:
- https://github.com/spring-projects/spring-data-couchbase/issues/1072
- https://github.com/spring-projects/spring-data-couchbase/issues/1066
Would be OK if I create a PR against 4.1.x for review?

**Comment by mikereiche** [comment]
@mmonti - is it important/helpful to you to get quick fixes in 4.1.x? Or just before the next release (which is February 17)? We could merge a quick-fix now as long as we got in a proper fix before then.

For https://github.com/spring-projects/spring-data-couchbase/issues/1066 - the issue about always back-ticking the 'key' of a QueryCriteria - the fix would be to accept a N1qlExpression as the key of a QueryCriteria (as well as the current String) - and propagate that change all the way through to where the maybeBackTic() method is called (it would just be a toString() on the N1qlExpression). [the quick fix would be just to check for "META(" in the string, ignoring case]

For https://github.com/spring-projects/spring-data-couchbase/issues/1072 - The PartTree processing of META properties (id and cas and expiration). The catch is that an entity could have a property named id, cas or expiration, that isn't the actual id, cas or expiration.   The id can be handled by isIdProperty(). but I don't have a solution for cas and expiration (probably add an \@CAS and \@Expiration field annotation, similar to \@id). [ the quick fix that would work for your case would just to do what I did with isIdProperty() ]
